### ISSUE TITLE
Odd bug introduced by the fix for #3057

### ISSUE THREAD

**Issue by taxilian** [Automattic/mongoose#4658]
Hey,

We just bisected an issue and found that it was introduced by commit d7ebe04a8ae521, the fix for #3057.  I'm really baffled how it could be causing issues, particularly this one.

It's in an awkward section of code where it's hard to tell what exactly is happening, but basically we save a document in one process and then load it in another and it's the wrong version.  I'm also not seeing __v get updated despite changing the document and saving it again.

I'll keep trying to track down what is going on, but I thought I'd let you know that there seems to be a problem caused by that change.  Let me know if you have any suggestions for what I should look at to figure out what is actually going on.

Richard

**Comment by taxilian** [comment]
I still can't tell exactly what it's doing, but here is what I had to change to make it work:

before: `let a = new Thingy({ref: MyDoc})`

after: `let a = new Thingy({ref: MyDoc._id})`

Previously we've always been able to assign a document (rather than the document's _id) to a field and have it work, but as of this commit at least when created as above it no longer works.

This kinda scares me because I don't know where else we might be relying on that syntax that won't work now. We're reverting to 4.6.4 for now.

**Comment by vkarpov15** [comment]
Hmm that's strange, thanks for reporting. I'll investigate asap but any extra information you may have would be very helpful.

**Comment by taxilian** [comment]
Further investigation shows that it seems to be related to saving the document, changing it, and then saving it again.

**Comment by vkarpov15** [comment]
I'm gonna need more info. The below script shows the behavior you mentioned, `child` is a doc rather than an objectid, but that's expected behavior and it works this way in 4.6.5, 4.6.0, and 4.5.0. Can you modify it to more closely match your use case?

```javascript
'use strict';

var assert = require('assert');
var mongoose = require('mongoose');
var Schema = mongoose.Schema;

mongoose.connect('mongodb://localhost/gh4658');
mongoose.set('debug', true);

const ChildSchema = new mongoose.Schema({
  name: String
});

const ChildModel = mongoose.model('Child', ChildSchema);

const ParentSchema = new mongoose.Schema({
  child: { type: Schema.Types.ObjectId, ref: 'Child' }
});

var ParentModel = mongoose.model('Parent', ParentSchema);

ChildModel.create({ name: 'Luke'}).
  then(child => {
    console.log('doc', new ParentModel({ child }));
    process.exit(0);
  }).
  catch(error => {
    console.error(error);
    process.exit(-1);
  });
```

**Comment by jloveridge** [comment]
The issue is actually very subtle and only seems to occur if updating the document with a populated field that is part of the models shard key. If you were to modify your example above to add `name: String` to the `ParentSchema` and then to add `{shardKey: {child: 1, _id: 1}` as schema options you would be able to experience the issue. What happens is the `update` query uses the populated form of the `child` attribute as part of the criteria. This obviously isn't going to match the `ObjectId` that exists at that path in the DB thus the document doesn't get updated. The call to `save` indicates success but the `update` doesn't actually update anything and subsequently loading the document results in it being in the original state. Here is a modified version of your code that reproduces the issue. (Note: the update works correctly in mongoose 4.6.4 and earlier.)

```javascript
'use strict';

var mongoose = require('mongoose');
var Schema = mongoose.Schema;

mongoose.connect('mongodb://localhost/gh4658');
mongoose.set('debug', true);

const ChildSchema = new mongoose.Schema({
  name: String
});

const ChildModel = mongoose.model('Child', ChildSchema);

const ParentSchema = new mongoose.Schema({
  name: String,
  child: { type: Schema.Types.ObjectId, ref: 'Child' }
}, {shardKey: {child: 1, _id: 1}});

var ParentModel = mongoose.model('Parent', ParentSchema);

ChildModel.create({ name: 'Luke'}).
  then(child => {
    const parent = new ParentModel({name: 'Jane'});
    parent.child = child;
    return parent.save();
  }).
  then(parent => {
    console.log('parent:', parent);
    parent.name = 'Sam';
    return parent.save();
  }).
  then(parent => {
    if (parent.name !== 'Sam') {
      process.exit(-1);
    }
    // Now load the document from the DB
    return ParentModel.findById(parent._id);
  }).
  then(parent => {
    console.log('parent:', parent);
    if (parent.name == 'Sam') {
      process.exit(0);
    } else {
      process.exit(1);
    }
  }).
  catch(error => {
    console.error(error);
    process.exit(-1);
  });
```

**Comment by jloveridge** [comment]
To me the worst part about this bug is that the save appears to succeed thus the programmer would go on their merry way thinking that the document has been updated properly and only later discovers that the document is in the wrong state.

**Comment by vkarpov15** [comment]
Good catch @jloveridge , fixed in master, will be released with 4.6.7 :+1:

**Comment by jloveridge** [comment]
Great, thanks!
